test/test_distributed.py

```diff
@@ -484,21 +551,150 @@ def _test_barrier_helper(self, group, group_id, rank):
+        nGPUs = torch.cuda.device_count()
+        world_size = dist.get_world_size()
+        visible_devices = range(nGPUs)
+        nGPUs_per_process = int(nGPUs / world_size)
+        visible_gpus_per_rank = {}
+        for i in range(world_size):
+            visible_gpus_per_rank[i] = visible_devices[
+                i * nGPUs_per_process: (i + 1) * nGPUs_per_process]
+        return visible_gpus_per_rank
+
+    def _test_broadcast_multigpu_helper(self, group, group_id,
+                                        rank, visible_gpus_per_rank):
+        for src in group:
+            expected_tensor = _build_tensor(src + 1)
+            tensors = [_build_tensor(src + 1, -1).cuda(device=i)
+                       for i in visible_gpus_per_rank[rank]]
+            if rank == src:
+                tensors[0] = expected_tensor.cuda(
+                    device=visible_gpus_per_rank[rank][0])
+            dist.broadcast_multigpu(tensors, src, group_id)
+            for tensor in tensors:
+                self.assertEqual(tensor, expected_tensor)
+        self._barrier()
+
+    @unittest.skipIf(BACKEND != 'nccl',
+                     "Only Nccl backend supports broadcast multigpu")
+    @skip_if_no_cuda_distributed
+    @skip_if_single_gpu
+    def test_broadcast_multigpu(self):
+        group, group_id, rank = self._init_global_test()
+        visible_gpus_per_rank = self._init_multigpu_helper()
+        self._test_broadcast_multigpu_helper(group, group_id,
+                                             rank, visible_gpus_per_rank)
+
+    def _test_all_reduce_multigpu_helper(self, group, group_id, rank,
+                                         visible_gpus_per_rank, op,
+                                         master_value, worker_value,
+                                         expected_value):
+        for src in group:
+            if rank == src:
+                tensors = [_build_tensor(src + 1, master_value).cuda(device=i)
+                           for i in visible_gpus_per_rank[rank]]
+            else:
+                tensors = [_build_tensor(src + 1, worker_value).cuda(device=i)
+                           for i in visible_gpus_per_rank[rank]]
+
+            dist.all_reduce_multigpu(tensors, op, group_id)
+            expected_tensor = _build_tensor(src + 1, expected_value)
+            for tensor in tensors:
+                self.assertEqual(tensor, expected_tensor)
+
+        self._barrier()
+
+    @unittest.skipIf(BACKEND != 'nccl',
+                     "Only Nccl backend supports allreduce multigpu")
+    @skip_if_no_cuda_distributed
+    @skip_if_single_gpu
+    def test_all_reduce_multigpu(self):
+        group, group_id, rank = self._init_global_test()
+        visible_gpus_per_rank = self._init_multigpu_helper()
+        self._test_all_reduce_multigpu_helper(
+            group, group_id, rank, visible_gpus_per_rank, dist.reduce_op.SUM,
+            2, 10, (2 + 10 * (len(group) - 1)) * len(visible_gpus_per_rank[0]))
```

Comment:
this looks like it assumes that you have an equal number of GPUs for each rank, which might not be the case

Reply:
I think the two wrappers(@skip_if_no_cuda_distributed,   @skip_if_single_gpu) ensures it skips testing if there is not cuda device/ #device <2. These two wrappers comes with all multigpu testcases. Please correct me if this is not enough :)

Reply:
For the equal number per process, this is actually enforced by current Nccl API. I talked to @teng-li about this. It's caused by how the nccl communicator was created, (gpu0 on node0, gpu0 on node1,...) is in one nccl comm group. 
I also tested it with one process has gpu0 and the other process has gpu1,2, it fails. 
IMHO we need to keep the equal number of gpu per process until this is changed in Nccl backend.

Reply:
Sorry, can you please explain what do you mean by `cuda device/ #device <2`? If the number of GPUs in the system is odd we should either skip some of them, or skip these tests instead of making it a hard failure (alternatively we can run with `WORLD_SIZE` equal to the number of GPUs)